### ISSUE TITLE
Function parsing throws an assert

### ISSUE THREAD

**Issue by natashenka** [Microsoft/ChakraCore#3046]
The following script (a v8 test case) throws an assert when I run it with Chakra:

load("test/mjsunit/wasm/wasm-constants.js");
load("test/mjsunit/wasm/wasm-module-builder.js");

(function() {
"use asm";
  var builder = new WasmModuleBuilder();
  builder.addFunction("regression_702460", kSig_i_v)
    .addBody([
        kExprI32Const, 0x52,
        kExprI32Const, 0x41,
        kExprI32Const, 0x3c,
        kExprI32Const, 0xdc, 0x01,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprSetLocal, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprGrowMemory, 0x00,
        kExprS128LoadMem, 0x00, 0x40,
        kExprUnreachable,
        kExprGrowMemory, 0x00
        ]).exportFunc();
  assertThrows(() => builder.instantiate());
})();

It would be preferable if this did not throw an assert unless it puts Chakra in a bad state, as it makes testing more difficult

**Comment by Cellule** [comment]
Do you mean it asserts in chakra or in the test runner ?
I tried running it with ch.exe and I get the following output
```bash
ch runner.js -args regress\wasm\regression-702460.js -endargs
Error: Expected () => builder.instantiate() to throw
```
where runner.js sets the `load` and `assertThrows` function.

However, if I run with 
```bash
 ch runner.js -args regress\wasm\regression-702460.js -endargs -off:wasmdeferred
Expected Error: WebAssemblyCompileError: function regression_702460[0] at offset 12/67: unknown memory
```

Chakra correctly throws an exception when trying to use the grow_memory operator when no memory is specified in the module.
We lazily evaluate functions which is why we don't throw at instantiation time. The flag `-off:wasmdeferred` disables that feature and pre-parses the whole module.

If you did encounter an Assert in Chakra, could you paste what the error is ?

**Comment by natashenka** [comment]
Just in Chakra. I'm running:

ChakraCore/out/Debug/ch v8/test/mjsunit/regress/wasm/regression-702460.js

I'm using the Linux debug build (should have mentioned that before). The assert is:

ASSERTION 21059: (/home/user/test_everywhere/ChakraCore/lib/Parser/Parse.cpp, line 5247) Deferring function parsing of a function does not handle lambda syntax
 Failure: (!fLambda)
Illegal instruction (core dumped)

**Comment by Cellule** [comment]
Oh I see the error. This was fixed in #3021
We don't support lambda in asm.js module. 
In this particular test case, I don't understand why there is a `"use asm"` statement since it is not an asm.js module. 
However, it should be fixed with latest build. We'll say it's an invalid asm.js module and parse as normal javascript.

Take note however, that we don't fully support wasm on linux right now (still some issues with floats).

**Comment by natashenka** [comment]
I pulled the new code, and it doesn't assert anymore. Thanks!
